File: app/src/main/java/com/monk/aidldemo/binder/messenger/MyMessengerService.java
package com.monk.aidldemo.binder.messenger;

import android.app.Service;
import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.os.IBinder;
import android.os.Message;
import android.os.Messenger;
import android.os.RemoteException;

import com.monk.commonutils.LogUtil;

import java.lang.ref.WeakReference;

/**
 * @author monk
 * @date 2019-01-15
 */
public class MyMessengerService extends Service {

    private Messenger messenger = new Messenger(new MessengerHandler(this));

    private static class MessengerHandler extends Handler {
        WeakReference<Service> weakReference;

        MessengerHandler(Service service) {
            weakReference = new WeakReference<>(service);
        }

        @Override
        public void handleMessage(Message msg) {
            if (weakReference.get() != null) {
                LogUtil.v("MyMessengerService", msg.getData().getString("msg"));
                Messenger client = msg.replyTo;
                Message serverMsg = Message.obtain(null, 0);
                Bundle bundle = new Bundle();
                bundle.putString("msgServer", "服务器返回数据");
                serverMsg.setData(bundle);
                try {
                    client.send(serverMsg);
                } catch (RemoteException e) {
                    e.printStackTrace();
                }
            }
        }

    }

    @Override
    public IBinder onBind(Intent intent) {
        return messenger.getBinder();
    }
}

File: module_fragment/src/main/java/com/monk/modulefragment/aidl/MyMessengerService.kt
package com.monk.modulefragment.aidl

import android.app.Service
import android.content.Intent
import android.os.*
import com.monk.commonutils.LogUtil
import java.lang.ref.WeakReference

/**
 * @author monk
 * @date 2019-01-15
 */
class MyMessengerService : Service() {
    private val messenger = Messenger(MessengerHandler(this))

    private class MessengerHandler internal constructor(service: Service?) : Handler() {
        var weakReference: WeakReference<Service?> = WeakReference(service)
        override fun handleMessage(msg: Message) {
            if (weakReference.get() != null) {
                LogUtil.v("MyMessengerService", msg.data.getString("msg"))
                val client = msg.replyTo
                val serverMsg = Message.obtain(null, 0)
                val bundle = Bundle()
                bundle.putString("msgServer", "服务器返回数据")
                serverMsg.data = bundle
                try {
                    client.send(serverMsg)
                } catch (e: RemoteException) {
                    e.printStackTrace()
                }
            }
        }

    }

    override fun onBind(intent: Intent): IBinder? {
        return messenger.binder
    }
}